src/topologies/halfedge.jl

```diff
@@ -144,126 +144,119 @@ function HalfEdgeTopology(elems::AbstractVector{<:Connectivity}; sort=true)
+    u = inds[i]
+    v = inds[mod1(i + 1, n)]
+    # insert halfedges u -> v and v -> u
+    he = get!(() -> HalfEdge(u, elem), half4pair, (u, v))
+    half = get!(() -> HalfEdge(v, nothing), half4pair, (v, u))
+    he.half = half
+    half.half = he
   end
 
-  # insert all other elements
-  for e in 2:length(adjelems)
-    elem = adjelems[e]
-    inds = collect(indices(elem))
-    v = CircularVector(inds)
-    n = length(v)
-    for i in 1:n
-      # if pair of vertices is already in the
-      # dictionary this means that the current
-      # polygon has inconsistent orientation
-      if haskey(half4pair, (v[i], v[i + 1]))
-        # delete inserted pairs so far
-        CCW[e] = false
-        for j in 1:(i - 1)
-          delete!(half4pair, (v[j], v[j + 1]))
-        end
-        break
+  # update half-edges using all other elements
+  added = false
+  disconnected = false
+  remaining = collect(2:length(adjelems))
+  while !isempty(remaining)
+    iter = 1
+    while iter ≤ length(remaining)
+      other = remaining[iter]
+      elem = eleminds[other]
+      inds = adjelems[other]
+      n = length(inds)
+      if anyhalf(inds, half4pair) || disconnected
+        # at least one half-edge has been found, so we can assess
+        # the orientation w.r.t. previously added elements/edges
+        added = true
+        disconnected = false
+        deleteat!(remaining, iter)
       else
-        # insert pair in consistent orientation
-        half4pair[(v[i], v[i + 1])] = HalfEdge(v[i], eleminds[e])
+        iter += 1
+        continue
       end
-    end
 
-    if !CCW[e]
-      # reinsert pairs in CCW orientation
-      for i in 1:n
-        half4pair[(v[i + 1], v[i])] = HalfEdge(v[i + 1], eleminds[e])
+      # if the other element is already claimed by any half-edge
+      # then the element must be reversed before further updates
+      isreversed[other] = anyhalfclaimed(inds, half4pair)
+
+      # insert half-edges in consistent orientation
+      if isreversed[other]
+        for i in eachindex(inds)
+          u = inds[i]
+          v = inds[mod1(i + 1, n)]
+          he = get!(() -> HalfEdge(v, elem), half4pair, (v, u))
+          if isnothing(he.elem)
+            he.elem = elem
+          else
+            assertion(
+              he.elem === elem,
+              lazy"duplicate edge $((v, u)) for element $(elem) is inconsistent with previous edge $he"
+            )
+          end
+          half = get!(() -> HalfEdge(u, nothing), half4pair, (u, v))
+          he.half = half
+          half.half = he
+        end
+      else
+        for i in eachindex(inds)
+          u = inds[i]
+          v = inds[mod1(i + 1, n)]
+          he = get!(() -> HalfEdge(u, elem), half4pair, (u, v))
+          he.elem = elem # this may be a pre-allocated half-edge with a nothing `elem`
+          half = get!(() -> HalfEdge(v, nothing), half4pair, (v, u))
+          he.half = half
+          half.half = he
+        end
       end
```

Comment:
In the meantime, I wonder if we could collapse these two branches into a single one where `inds` are looped in reverse order when `reversed[other]` is true. I see that the code is pretty much the same, except that the first branch has an additional `isnothing` check and an `assertion`. If you feel that we can't simplify this further, that is fine too. Just trying to keep things as simple as possible for future maintenance.